hikari/api/special_endpoints.py

```diff
@@ -1869,11 +1894,19 @@ def add_button(
         """
 
     @abc.abstractmethod
-    def add_select_menu(self: _T, custom_id: str, /) -> SelectMenuBuilder[_T]:
+    def add_select_menu(
+        self: _T,
+        type_: typing.Union[components_.ComponentType, int, str],
```

Comment:
Might want to use overloads here to keep the deprecated signature separate type-checking wise (and also in the reprs ppl get for the method in IDEs).

Reply:
Unfortunately there is no way to do this properly with overloads, as the signatures conflict

Reply:
Wdym, would it not just be 
```py
@typing.overload
def add_select_menu(self: _T, custom_id: str, /) -> SelectMenuBuilder[_T]:
    ...

@typing.overload
def add_select_menu(
    self: _T, type_: typing.Union[components_.ComponentType, int], custom_id: str, /
) -> SelectMenuBuilder[_T]:
    ...

def add_select_menu(
    self: _T,
    custom_id_or_type: typing.Union[components_.ComponentType, int, str],
    # These have default during the deprecation period for backwards compatibility, as custom_id
    # used to come first
    custom_id: str = "",
    /,
) -> SelectMenuBuilder[_T]:
    ...
```

Reply:
Would also be helpful to have overloads which give you more concise type based return types so that might end up looking more like
```py
@typing.overload  # Deprecated overload
@abc.abstractmethod
def add_select_menu(self: _T, custom_id: str, /) -> SelectMenuBuilder[_T]:
    ...

@typing.overload
@abc.abstractmethod
def add_select_menu(
    self: _T,
    type_: typing.Literal[components_.ComponentType.TEXT_SELECT_MENU, 3],
    custom_id: str,
    /,
) -> TextSelectMenuBuilder[_T]:
    ...

@typing.overload
@abc.abstractmethod
def add_select_menu(
    self: _T,
    type_: typing.Literal[components_.ComponentType.CHANNEL_SELECT_MENU, 8],
    custom_id: str,
    /,
) -> ChannelSelectMenuBuilder[_T]:
    ...

@typing.overload
@abc.abstractmethod
def add_select_menu(
    self: _T, type_: typing.Literal[components_.ComponentType, int], custom_id: str, /
) -> SelectMenuBuilder[_T]:
    ...

@abc.abstractmethod
def add_select_menu(
    self: _T,
    custom_id_or_type: typing.Union[components_.ComponentType, int, str],
    # These have default during the deprecation period for backwards compatibility, as custom_id
    # used to come first
    custom_id: str = "",
    /,
) -> SelectMenuBuilder[_T]:
    ...
```

Reply:
It's a shame typing-extensions hasn't done a release recently cause would've been helpful to annotate  the deprecated signature as `@typing_extensions.deprecated("message")` for type-checkers

Reply:
I have no idea what i was doing wrong, but overloads work now, added! Thanks!

And yeah, not having `@typing_extensions.deprecated("message")` is a bummer. I found out about the PEP yesterday and really hoped i could use it